dali/base/dautils.cpp

```diff
@@ -3083,3 +3113,262 @@ bool traceAllTransactions(bool on)
+            lock->setProp(getLockDataFieldName(LDFXPath), xPath.get());
+
+            unsigned epCount;
+            mb.read(epCount);
+            lock->setPropInt(getLockDataFieldName(LDFEPCount), epCount);
+            if (epCount < 1)
+                return lock;
+
+            for (unsigned l = 0; l < epCount; l++)
+            {
+                StringAttr ep;
+                __int64 connId, sessId;
+                unsigned mode, timeLockObtained;
+                mb.read(ep);
+                mb.read(sessId);
+                mb.read(connId);
+                mb.read(mode);
+                mb.read(timeLockObtained);
+
+                IPropertyTree* sub = createPTree("EP");
+                sub->setProp(getLockDataFieldName(LDFEP), ep.get());
+                sub->setPropInt64(getLockDataFieldName(LDFconnId), connId);
+                sub->setPropInt64(getLockDataFieldName(LDFsessId), sessId);
+                sub->setPropInt(getLockDataFieldName(LDFmode), mode);
+                sub->setPropInt(getLockDataFieldName(LDFtimeLocked), timeLockObtained);
+                lock->addPropTree("EP", sub);
+            }
+            return lock;
+        }
+
+    public:
+        IMPLEMENT_IINTERFACE;
+        MemoryBuffer mb;
+        unsigned lockCount;
+
+        bool first()
+        {
+            mb.reset();
+            mb.read(lockCount);
+
+            return next();
+        }
+
+        bool next()
+        {
+            cur.clear();
+            if (mb.getPos()>=mb.length())
+                return false;
+            cur.setown(deserializeLockData(mb));
+            return true;
+        }
+
+        bool isValid()
+        {
+            return cur.get()!=NULL;
+        }
+
+        IPropertyTree  & query()
+        {
+            return *cur;
+        }
+    } *fai = new CLockDataIterator;
+    mb.swapWith(fai->mb);
+    return fai;
+}
+
+IPropertyTreeIterator *CLockDataHelper::getLockDataTreeIterator(MemoryBuffer &lockInfo)
+{
+    return deserializeLockDataIterator(lockInfo);
+}
+
+StringBuffer &CLockDataHelper::formatLockData(MemoryBuffer &lockInfo, StringBuffer &out)
+{
+    StringAttr xPath;
+    lockInfo.read(xPath);
+    out.append("Locks on path: /").append(xPath).newline();
+    out.append("Endpoint            |SessionId       |ConnectionId    |mode    |time(duration)]").newline().newline();
+
+    unsigned lockCount;
+    lockInfo.read(lockCount);
+    if (lockCount < 1)
+    {
+        out.newline();
+        return out;
+    }
+
+    unsigned msNow = msTick();
+    CDateTime time;
+    time.setNow();
+
+    for (unsigned l = 0; l < lockCount; l++)
+    {
+        StringAttr ep;
+        __int64 connId, sessId;
+        unsigned mode, timeLockObtained;
+        lockInfo.read(ep);
+        lockInfo.read(sessId);
+        lockInfo.read(connId);
+        lockInfo.read(mode);
+        lockInfo.read(timeLockObtained);
+
+        CDateTime timeLocked;
+        unsigned lockedFor = msNow-timeLockObtained;
```

Comment:
Minor: Should possibly calculate and serialize the timelocked//duration at serialization time, leaving the formatting to print the info only.
As time will have elapsed between request, serialization and formatting.